From 6740f7b6fde099e8436841cf9a2ee42c16f3e516 Mon Sep 17 00:00:00 2001
From: Erwin Pannecoucke <erwin.pannecoucke@gmail.com>
Date: Sat, 18 Aug 2012 10:56:59 +0200
Subject: [PATCH] Fix first frame on rotation animation for msm7k

Modify rotation animation to use a screenshot instead of a black screen.

Original Author: Shane Francis
Original commit: http://review.cyanogenmod.com/#/c/18538/1

Change-Id: I628c0ffb36bd822d46d4bb039cd741f6d4fae88a
---
 .../android/server/wm/ScreenRotationAnimation.java |   39 ++++++++++++++++++--
 1 file changed, 35 insertions(+), 4 deletions(-)

diff --git a/services/java/com/android/server/wm/ScreenRotationAnimation.java b/services/java/com/android/server/wm/ScreenRotationAnimation.java
index 7e3261b..33f4380 100644
--- a/services/java/com/android/server/wm/ScreenRotationAnimation.java
+++ b/services/java/com/android/server/wm/ScreenRotationAnimation.java
@@ -188,7 +188,9 @@ class ScreenRotationAnimation {
 
     public ScreenRotationAnimation(Context context, SurfaceSession session,
             boolean inTransaction, int originalWidth, int originalHeight, int originalRotation) {
+        boolean isMsm7k = android.os.SystemProperties.get("ro.board.platform","").equals("msm7k");
         mContext = context;
+
         // Allow for abnormal hardware orientation
         mSnapshotRotation = (4 - android.os.SystemProperties.getInt("ro.sf.hwrotation",0) / 90) % 4;
         if (mSnapshotRotation == Surface.ROTATION_0 || mSnapshotRotation == Surface.ROTATION_180) {
@@ -223,11 +225,11 @@ class ScreenRotationAnimation {
         try {
             try {
                 if (WindowManagerService.DEBUG_SURFACE_TRACE) {
-                    mSurface = new SurfaceTrace(session, 0, "FreezeSurface", -1, mWidth, mHeight,
-                        PixelFormat.OPAQUE, Surface.FX_SURFACE_SCREENSHOT | Surface.HIDDEN);
+                    mSurface = new SurfaceTrace(session, 0, "FreezeSurface", 
+                        -1, mWidth, mHeight, PixelFormat.OPAQUE, isMsm7k ? 0 : (Surface.FX_SURFACE_SCREENSHOT | Surface.HIDDEN));
                 } else {
-                    mSurface = new Surface(session, 0, "FreezeSurface", -1, mWidth, mHeight,
-                        PixelFormat.OPAQUE, Surface.FX_SURFACE_SCREENSHOT | Surface.HIDDEN);
+                    mSurface = new Surface(session, 0, "FreezeSurface", 
+                        -1, mWidth, mHeight, PixelFormat.OPAQUE, isMsm7k ? 0 : (Surface.FX_SURFACE_SCREENSHOT | Surface.HIDDEN));
                 }
                 if (!mSurface.isValid()) {
                     // Screenshot failed, punt.
@@ -246,6 +248,35 @@ class ScreenRotationAnimation {
                             "  FREEZE " + mSurface + ": CREATE");
 
             setRotation(originalRotation);
+
+            if (isMsm7k) {
+                Rect rect = new Rect(0, 0, mWidth, mHeight);
+                Canvas canvas = null;
+
+                try {
+                    canvas = mSurface.lockCanvas(rect);
+                } catch (IllegalArgumentException e) {
+                    Slog.w(TAG, "Unable to lock surface", e);
+                } catch (Surface.OutOfResourcesException e) {
+                    Slog.w(TAG, "Unable to lock surface", e);
+                }
+
+                Bitmap screenshot = Surface.screenshot(0, 0);
+                if (canvas == null || screenshot == null) {
+                    Slog.w(TAG, "Null surface canvas");
+                    mSurface.destroy();
+                    mSurface = null;
+                    return;
+                }
+
+                Paint paint = new Paint(0);
+                paint.setXfermode(new PorterDuffXfermode(PorterDuff.Mode.SRC));
+
+                canvas.drawBitmap(screenshot, 0, 0, paint);
+                mSurface.unlockCanvasAndPost(canvas);
+
+            }
+
         } finally {
             if (!inTransaction) {
                 Surface.closeTransaction();
-- 
1.7.9.5

